core/drivers/crypto/stm32/stm32_hash.c

```diff
@@ -0,0 +1,909 @@
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	/*
+	 * The queue size is block_size + one register at first
+	 * then block_size.
+	 * So we may need to save at max queue_size + 3 bytes.
+	 * Let allocate a number of uin32_t: queue_size + 4.
+	 */
+	c->remain.buf = calloc(c->block_size + sizeof(uint32_t), 1);
+	if (!c->remain.buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	get_save_registers(c, &nb_reg, &first, &hmac_nb_reg, &hmac_first);
+
+	c->csr = calloc(nb_reg + hmac_nb_reg, sizeof(uint32_t));
+	if (!c->csr) {
+		free(c->remain.buf);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	return TEE_SUCCESS;
+}
+
+void stm32_hash_free(struct stm32_hash_context *c)
+{
+	if (!c)
+		return;
+
+	free(c->remain.buf);
+	free(c->csr);
+}
+
+TEE_Result stm32_hash_update(struct stm32_hash_context *c,
+			     const uint8_t *buffer, size_t len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t next_queue_size = c->queue_size;
+	vaddr_t base = 0;
+
+	assert(c);
+
+	base = c->dev->pdata.base;
+
+	if (!len || !buffer)
+		return TEE_SUCCESS;
+
+	mutex_lock(&c->dev->lock);
+	if (clk_enable(c->dev->pdata.clock)) {
+		EMSG("Fail to enable clk %s", c->dev->pdata.clock->name);
+		panic();
+	}
+
+	res = restore_context(c);
+	if (res)
+		goto exit;
+
+	/* We cannot fill the fifo */
+	if (c->remain.len + len < c->queue_size) {
+		if (!c->remain.buf) {
+			res = TEE_ERROR_BAD_STATE;
+			goto exit;
+		}
+
+		memcpy(((uint8_t *)c->remain.buf) + c->remain.len, buffer,
+		       len);
+		c->remain.len += len;
+
+		next_queue_size = c->queue_size;
+
+		/*
+		 * We don't need to save status as we didn't change IP
+		 * internal state.
+		 */
+		goto exit;
+	} else {
+		next_queue_size = c->block_size;
+	}
+
+	/*
+	 * First write data saved in previous update
+	 */
+	if (c->remain.len) {
+		size_t align = 0;
+		size_t i = 0;
+
+		if (!c->remain.buf) {
+			res = TEE_ERROR_BAD_STATE;
+			goto exit;
+		}
+
+		/* Add bytes needed to align saved data */
+		align = ROUNDUP(c->remain.len, sizeof(uint32_t)) -
+			c->remain.len;
+		memcpy(((uint8_t *)c->remain.buf) + c->remain.len, buffer,
+		       align);
+		c->remain.len += align;
+		buffer += align;
+		len -= align;
+
+		for (i = 0; i < c->remain.len / sizeof(uint32_t); i++) {
+			res = hash_write_data(base, c->remain.buf[i]);
+			if (res)
+				goto exit;
+
+			c->remain.buf[i] = 0; /* Reset to 0 */
+		}
+
+		/* No more saved data */
+		c->remain.len = 0;
+	}
+
+	/*
+	 * Here, the data should be written to the FIFO until we cannot
+	 * guarantee anymore that the data that we write will trigger a
+	 * process of data. Then we write the remaining data until DINIS
+	 * is set to 1 by hardware, meaning that a complete block can be
+	 * sent. Data written will be saved during save_context() and
+	 * remaining data not written (if there's any) will be saved in
+	 * c->remain.buf.
+	 */
+	while (len >= c->queue_size ||
+	       !(io_read32(base + _HASH_SR) & _HASH_SR_DINIS)) {
+		uint32_t tmp_buf = 0;
+
+		memcpy(&tmp_buf, buffer, sizeof(uint32_t));
```

Comment:
should `sizeof(uint32_t)` be replaced with `MIN(len ,sizeof(uint32_t))`.
Or maybe it is the `||` test at line 661 that should be a `&&`.
I must say that the first sentence in inline comment at line 653 is a bit cryptic to me.